consensus/core/src/dag_state.rs

```diff
@@ -755,6 +739,67 @@ impl DagState {
         blocks.first().cloned().unwrap()
     }
 
+    // Sets the block as committed in the cache. If the block is set as committed for first time, then true is returned, otherwise false is returned instead.
+    // Method will panic if the block is not found in the cache.
+    pub(crate) fn set_committed(&mut self, block_ref: &BlockRef) -> bool {
+        if let Some(block_info) = self.recent_blocks.get_mut(block_ref) {
+            if !block_info.committed {
+                block_info.committed = true;
+                return true;
+            }
+            false
+        } else {
+            panic!(
+                "Block {:?} not found in cache to set as committed.",
+                block_ref
+            );
+        }
+    }
+
+    /// Returns true if the block is committed. Only valid for blocks above the GC round.
+    pub(crate) fn is_committed(&self, block_ref: &BlockRef) -> bool {
+        self.recent_blocks
+            .get(block_ref)
+            .unwrap_or_else(|| panic!("Attempted to query for commit status for a block not in cached data {block_ref}"))
+            .committed
+    }
+
+    /// Recursively sets blocks in the causal history of the root block as hard linked, including the root block itself.
+    /// Returns the list of blocks that are newly linked.
+    /// The returned blocks are guaranteed to be above the GC round.
+    /// NOTE: this function should only be called with GC enabled.
+    pub(crate) fn recursive_hard_link(&mut self, root_block: BlockRef) -> Vec<BlockRef> {
```

Comment:
can you add a unit test for this in dag state, unless you think its thoroughly tested elsewhere